### ISSUE TITLE
[Bug] Layer flickering when used with Redux

### ISSUE THREAD

**Issue by argilya** [visgl/deck.gl#7478]
### Description

Hello,

I have been dealing with a classic "flicker" bug in my Deck.gl application. During state transitions, controlled by `Redux`, an incorrect UI state is rendered for a single frame, creating an instantaneous flicker.

Reduced to the bare minimum, the application renders a grid of squares in a `ColumnLayer`. Each square has an id associated with it, and a pair of coordiates. When you click on a square, the id of the square gets dispatched to `Redux`. A change in `Redux` then triggers a `useEffect` that updates the `viewState` to the coordinates of the selected square.

When the transition of `viewState` begins, there is a flicker. Here is what it looks like in the the parent application running on my local:

https://user-images.githubusercontent.com/100630262/205774582-dde0e69a-244a-40e8-a613-2d58d2db4ae9.mp4

I re-created the basic code flow in [this sandbox](https://codesandbox.io/s/ghost-frame-bug-redux-only-mr8ioy?file=/src/App.jsx), and the bug occurs there too, but at a much smaller rate, eg every 1 in 10 to 20 transitions. Example from sandbox:

https://user-images.githubusercontent.com/100630262/205774826-3551484b-95db-4b04-aa3b-4977b888ff83.mp4

I found that this bug only occurs when two things are true together:
1) I use Deck's `viewState` and `onViewStateChange` props (if I use `initialViewState`, there is no flicker)
2) Redux is involved as an intermediary for triggering `viewState` updates (if I use a local `useState` hook to keep track of selected square id, there is no flicker)

### Flavors

- [X] React
- [ ] Python/Jupyter notebook
- [ ] MapboxLayer
- [ ] GoogleMapsOverlay
- [ ] CartoLayer
- [ ] DeckLayer/DeckRenderer for ArcGIS

### Expected Behavior

I expect there to be no erroneous frames, when I click on a square and trigger a view state update.

### Steps to Reproduce

Open the sandbox https://codesandbox.io/s/ghost-frame-bug-redux-only-mr8ioy?file=/src/App.jsx

Click on any of the black squares. About 10% of the time it will cause an incorrect frame render. After the initial click that zooms you in, the erroneous frame seems to occur more frequently if you zoom out and move your camera around before clicking on a square.

(In the parent app, running on local, the erroneous frame occurs on nearly all square transitions. I have not been able to reproduce the same bad render rate in the sandbox.)

### Environment

- Framework version (in sandbox):
  - `deck.gl 8.8.20`
  - `react 18.2.0`
  - `react-redux 8.0.5`
  - `redux 4.2.0`
  - `@reduxjs/toolkit 1.9.1`
- Browser: not sure what sandbox is using (on local I use Chrome 108)
- OS: not sure what sandbox is using (on local use Ubuntu 20.04)

### Logs

There are no logs or warnings.

**Comment by argilya** [comment]
Hello,

I am still struggling with this bug. I've been trying to develop more intuition on what's going here. Below is a summary of my new "findings":

### The flicker frame can be reproduced consistently in sandbox
>  I have not been able to reproduce the same bad render rate in the sandbox.

I found a way to consistently reproduce the flicker frame in the sandbox by increasing the number of items in the layer. I upped the number of squares from 100 to 1 million. So now, when you zoom out, then click on a square, you get the flicker virtually every time. Example (select new square twice, after zooming out, there is a flicker both times):

https://user-images.githubusercontent.com/100630262/210413778-80ff70bc-7e84-45f1-99e5-5710e0631249.mp4

Here is the sandbox:
https://codesandbox.io/s/ghost-frame-bug-redux-only-mr8ioy?file=/src/App.jsx

### The flicker happens with useState too

> Redux is involved as an intermediary for triggering viewState updates (if I use a local useState hook to keep track of selected square id, there is no flicker)

With 1 million items, it no longer matters if I use Redux or useState to keep track of selected item id. The flicker occurs both with Redux and with useState.

### The flicker is, in fact, an extra frame
![image](https://user-images.githubusercontent.com/100630262/210415161-b0938e6a-538a-4712-8890-f2b5a563ea6e.png)

Here is a gif of me flipping through the frames, with the emphasis on the flicker frame:

https://user-images.githubusercontent.com/100630262/210415397-61a02f11-80c3-4704-b1d2-34641ee5a9fe.mp4

### The bug can be worked around by triggered transitions in the onClick callback
One way I found to solve this bug is to disconnect view state transition from selected item id. Instead of setting new view state in a useEffect that listens for selected item id, I set the view state when I click on the new square. That removes the extra frame. See gif:

https://user-images.githubusercontent.com/100630262/210418879-cec12df0-bfa5-4c8d-af20-c2365eec60d2.mp4

Here is the sandbox with this approach:

https://codesandbox.io/s/ghost-frame-bug-redux-only-setviewstate-in-onclick-gbiqwm?file=/src/App.jsx

Is this the correct approach? That's how they trigger viewState transitions based on form input in this deck.gl [example](https://deck.gl/examples/multi-view/). Doing this in my parent app would require a messy refactor, so I am keen on figuring out what's wrong with my current `useEffect` based approach. 

### Some thoughts

I feel like I am missing something basic about how deck.gl works, but I don't see an immediate problem with my current bugged approach:
1) Render the layer of 1M items
2) Click on item, set item id in Redux
3) Redux hook triggers a re-render of the container component, causing the data object to be remade
4) Data object is passed into `DeckMap` component, the layer is re-rendered with new data. Then, a useEffect in the same component transitions to new viewState.

Are the two things in step 4 conflicting with each other (the re-render of new layer, and the useEffect)? But then, why is the conflict so inconsistent (eg if we render only a few items instead of a million, the flicker frame rarely occurs).

I would appreciate any comments. Thanks you!

PS
My bug is very similar to this one (which was successfully resolved), but I have not been able to gain insight from the discussion:

https://github.com/visgl/deck.gl/issues/4550

**Comment by argilya** [comment]
Hello,

I am going to close this ticket, since this issue is not a DeckGL bug.

For those who stumble on this thread in the future: I solved the flicker by refactoring how we manage state in our application. I cut all of the useEffects that connected UI interactions to viewState (which was a local variable inside the Deck's parent component). I placed viewState directly in Redux, and changes to viewState are now dispatched directly from UI's `onClick` callbacks, rather than through useEffect intermediaries.

I think having a chain of expensive useEffects that triggered each other, and then finally updated viewState, was the root cause of the problem. I still don't have a good intuition of what the _actual_ problem was, though. JavaScript running calls out of order, because of all the useEffects? Hopefully, as I gain more experience, I can understand low level a bit more.

Closing the thread now, thank you for reading!
